### PR TITLE
Fixing Gemma checkpointing

### PR DESCRIPTION
**Pull request by SalmanMohammadi**
#### Context

- [ ] add a new feature
- [x] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

https://github.com/pytorch/torchtune/issues/1062

#### Changelog

Added custom logic in HF checkpointing for Gemma models to correctly load and save tied embedding weights.

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help.)

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add unit tests for any new functionality
- [x] update docstrings for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof 
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

I've added a test for loading and saving gemma checkpoints.

### CODE REVIEW

**Review by pbontrager — COMMENTED** [review]
The original reason we left this off from the state_dict was because it's redundant information that'll be overridden once the weights are tied. Did the user, in the original issue, confirm that the Transformers model functions incorrectly without the missing key?

Otherwise, if it is strictly necessary to add the redundant weight info, everything looks good, and thank you so much for putting up this PR. This PR follows the same pattern that we did with the Mistral Reward model, but based on the small changes needed to the hf_to_tune and tune_to_hf models to make this work, I think we should keep our eye on this for generalizing the checkpointer logic @joecummings .

**Comment by SalmanMohammadi** [comment]
> Did the user, in the original issue, confirm that the Transformers model functions incorrectly without the missing key?

Yep. I replicated the issue and found Transformers was complaining about missing `lm_head.weight`. The [Gemma implementation in Transformers](https://github.com/huggingface/transformers/blob/14ff5dd962c1bd0a4e3adaac347ba396d8df5add/src/transformers/models/gemma/modeling_gemma.py#L1042) has an explicit key for this tied weight. They reccomend using a [conversion script in the Gemma docs](https://huggingface.co/docs/transformers/model_doc/gemma) which adds `lm_head.weight` to the state dict.

> I think we should keep our eye on this for generalizing the checkpointer logic 

Yep - I was thinking this too.

File: torchtune/models/gemma/_convert_weights.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    def _permute(t, n_heads):
+        return (
+            t.view(n_heads, 2, head_dim // 2, dim)
+            .transpose(1, 2)
+            .reshape((head_dim * n_heads), dim)
+        )
+
+    for key, value in state_dict.items():
+        if (
+            "rotary_emb.inv_freq" not in key and "lm_head.weight" not in key
```
[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply torchtune/models/gemma/_convert_weights.py:50]: If by "general" you mean torchtune, I believe so. I'm a little puzzled with how checkpointing was working before, e.g. in https://github.com/pytorch/torchtune/pull/668/files. I wrote the Gemma checkpointing tests first, and they were failing with the current checkpointing setup.

[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply torchtune/models/gemma/_convert_weights.py:50]: ^ I'll sanity check by loading this into a HF model just now.

**Comment by SalmanMohammadi** [comment]
@pbontrager looks all good
```python
import transformers
from transformers import GemmaForCausalLM, PretrainedConfig
hf_model = GemmaForCausalLM.from_pretrained("google/gemma-2b")
hf_state_dict = hf_model.state_dict()

from torchtune.models.gemma import gemma_2b
tune_model = gemma_2b()
tune_state_dict = tune_model.state_dict()

from torchtune.models.gemma import gemma_hf_to_tune, gemma_tune_to_hf
tune_state_dict_converted_to_hf = gemma_tune_to_hf(tune_state_dict)
hf_state_dict_convert_to_tune = gemma_hf_to_tune(hf_state_dict)

print(tune_model.load_state_dict(hf_state_dict_convert_to_tune))
# <All keys matched successfully>

print(hf_model.load_state_dict(tune_state_dict_converted_to_hf))
# <All keys matched successfully>
```

File: torchtune/models/gemma/_convert_weights.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply torchtune/models/gemma/_convert_weights.py:50]: > If by "general" you mean HF models, I believe so. EDIT: re-reading this, if "general" is llama-style models, then also yes :)
> 
> I'm a little puzzled with how checkpointing was working before, e.g. in https://github.com/pytorch/torchtune/pull/668/files. I wrote the Gemma checkpointing tests first, and they were failing with the current checkpointing setup.

Illustrating this further, this is the checkpointing behaviour I posted below, but on main:
```python
import transformers
from transformers import GemmaForCausalLM, PretrainedConfig
hf_model = GemmaForCausalLM.from_pretrained("google/gemma-2b")
hf_state_dict = hf_model.state_dict()

from torchtune.models.gemma import gemma_2b
tune_model = gemma_2b()
tune_state_dict = tune_model.state_dict()

from torchtune.models.convert_weights import tune_to_hf, hf_to_tune
tune_state_dict_converted_to_hf = tune_to_hf(tune_state_dict, num_heads=8,num_kv_heads=1, dim=2048, head_dim=256)
hf_state_dict_convert_to_tune = hf_to_tune(hf_state_dict,num_heads=8,num_kv_heads=1, dim=2048, head_dim=256)

print(tune_model.load_state_dict(hf_state_dict_convert_to_tune))
"""
---------------------------------------------------------------------------

RuntimeError                              Traceback (most recent call last)

[<ipython-input-6-fe655bbc13e2>](https://localhost:8080/#) in <cell line: 1>()
----> 1 print(tune_model.load_state_dict(hf_state_dict_convert_to_tune))
      2 # <All keys matched successfully>

[/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py](https://localhost:8080/#) in load_state_dict(self, state_dict, strict, assign)
   2187 
   2188         if len(error_msgs) > 0:
-> 2189             raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
   2190                                self.__class__.__name__, "\n\t".join(error_msgs)))
   2191         return _IncompatibleKeys(missing_keys, unexpected_keys)

RuntimeError: Error(s) in loading state_dict for GemmaTransformerDecoder:
	Unexpected key(s) in state_dict: "output.weight".
"""

print(hf_model.load_state_dict(tune_state_dict_converted_to_hf))

"""
---------------------------------------------------------------------------

RuntimeError                              Traceback (most recent call last)

[<ipython-input-7-537d93467a7f>](https://localhost:8080/#) in <cell line: 1>()
----> 1 print(hf_model.load_state_dict(tune_state_dict_converted_to_hf))
      2 # <All keys matched successfully>

[/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py](https://localhost:8080/#) in load_state_dict(self, state_dict, strict, assign)
   2187 
   2188         if len(error_msgs) > 0:
-> 2189             raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
   2190                                self.__class__.__name__, "\n\t".join(error_msgs)))
   2191         return _IncompatibleKeys(missing_keys, unexpected_keys)

RuntimeError: Error(s) in loading state_dict for GemmaForCausalLM:
	Missing key(s) in state_dict: "lm_head.weight".
"""
```

**Review by pbontrager — APPROVED** [review]
I believe it might have worked before if strict=False was set for load_state_dict? Either way, I agree it's better supported this way.

**Comment by SalmanMohammadi** [comment]
Thanks so much for the review @pbontrager. Hopefully good to go with green CI.

**Pull request merged**